src/natcap/invest/delineateit/delineateit.py

```diff
@@ -601,6 +630,193 @@ def snap_points_to_nearest_stream(points_vector_path, stream_raster_path_band,
+    # conversions that users might do.
+    target_layer.CreateField(
+        ogr.FieldDefn('point_id', ogr.OFTInteger64))
+
+    # Add a feature to the layer for each point
+    target_layer.StartTransaction()
+    for idx, (x, y) in enumerate(pour_point_set):
+        geometry = ogr.Geometry(ogr.wkbPoint)
+        geometry.AddPoint(x, y)
+        feature = ogr.Feature(target_defn)
+        feature.SetGeometry(geometry)
+        feature.SetField('point_id', idx)
+        target_layer.CreateFeature(feature)
+    target_layer.CommitTransaction()
+
+    target_layer = None
+    target_vector = None
+
+
+def _find_raster_pour_points(flow_dir_raster_path_band, raster_info):
+    """
+    Memory-safe pour point calculation from a flow direction raster.
+    
+    Args:
+        flow_dir_raster_path_band (tuple): tuple of (raster path, band index)
+            indicating the flow direction raster to use.
+        raster_info (dict): value of ``pygeoprocessing.get_raster_info(
+            flow_dir_raster_path)``. Avoiding redoing this function call
+            since it's used in the calling function ``detect_pour_points``.
+
+    Returns:
+        set of (x, y) coordinate tuples of pour points, in the same coordinate
+        system as the input raster.
+    """
+
+    flow_dir_raster_path, band_index = flow_dir_raster_path_band
+    # Open the flow direction raster band
+    raster = gdal.OpenEx(flow_dir_raster_path, gdal.OF_RASTER)
+    if raster is None:
+        raise ValueError(
+            "Raster at %s could not be opened." % flow_dir_raster_path)
+    band = raster.GetRasterBand(band_index)
+
+    pour_point_set = set()
+    # Read in flow direction data and find pour points one block at a time
+    for block in pygeoprocessing.iterblocks((flow_dir_raster_path, band_index),
+                                            offset_only=True): 
+        # Expand each block by a 1-pixel-wide margin so that they overlap
+        # Keep track of which block edges are raster edges
+        block, edges = _expand_and_find_edges(block, raster_info['raster_size'])
+        flow_dir_block = band.ReadAsArray(**block)
+        new_pour_points = delineateit_core.calculate_pour_point_array(
+            flow_dir_block.astype(numpy.intc), 
+            edges, 
+            raster_info['nodata'][0])
+
+        # Add the offsets so that points from all blocks reference the same 
+        # coordinate system (such that the upper-left corner of the raster is 
+        # (0, 0), and each pixel is 1x1).
+        for x, y in new_pour_points:
+            pour_point_set.add((x + block['xoff'], y + block['yoff']))
+            
+    raster = None
+    band = None
+
+    # Return coordinates in the same coordinate system as the input raster
+    origin = (raster_info['geotransform'][0], raster_info['geotransform'][3])
+    return _convert_numpy_coords_to_geotransform(
+        pour_point_set, origin, raster_info['pixel_size'])
```

Comment:
Good idea! I guess I was thinking to keep the cython function isolated from the context of each block, but there's no reason it needs to be that way, and this will be more efficient.